types/ember__object/evented.d.ts

```diff
@@ -51,6 +51,5 @@ export default Evented;
  * a specified event or events are triggered.
  */
 export function on(
-    eventNames: string,
-    func: (...args: any[]) => void
+    ...args: Array<string | ((...args: any[]) => void)>
```

Comment:
I believe this type is wrong, because it's not actually possible for these to be mixed. The type as written here would let you write this:

```ts
on(
  'click',
  (someEv) => {
    console.log(someEv);
  },
  'dblclick',
  'mouseover'
);
```

In order to make it take a spread of arguments *followed* by the callback, we should support it *now* via overloads, and then once we bump our minimum supported TS version to 4.0 (presumably in the first half 2021) I *think* we'll be able to use TS 4.0's support for variadic tuples to make this work as desired.

Reply:
Yeah, I agree that this definition is too accepting. I just chose it because it was what's in the source and the existing definition disallows valid options.

Reply:
@chriskrycho what do you have in mind for overloads? Just a number of variants with increasing numbers of eventName arguments?

Reply:
Also, I'm not seeing how variadic tuples would help us here, but it's also very possible that I'm missing something.

Reply:
Ah, looks like this would work in TS 4:

```ts
type EventNames = string[];
type OnArgs = [...EventNames, (...args: any[]) => void];
export function on(...args: OnArgs): (...args: any[]) => void;
```

Reply:
Exactly that for TS 4. For TS <4, yeah, you’d do a bunch of increasing event names—we tend to go up to 5 or 6 most similar places in the Ember types.